workflow/function-cacher/index.js

```diff
@@ -0,0 +1,72 @@
+math.config({
+  number: 'BigNumber',
+});
+
+const deepEqual = (firstObject, secondObject) => {
+  if (firstObject === secondObject) {
+    return true;
+  } else if (
+    typeof firstObject === 'object' &&
+    firstObject !== null &&
+    typeof secondObject === 'object' &&
+    secondObject !== null
+  ) {
+    if (Object.keys(firstObject).length !== Object.keys(secondObject).length) {
+      return false;
+    }
+
+    for (let key in firstObject) {
+      if (secondObject.hasOwnProperty(key)) {
+        if (!deepEqual(firstObject[key], secondObject[key])) {
+          return false;
+        }
+      } else {
+        return false;
+      }
+    }
+
+    return true;
+  } else {
+    return false;
+  }
+};
+
+class Cacher {
+  constructor() {}
+
+  withCache(callback) {
+    let cache = {};
+
+    return (...arg) => {
+      for (let key in cache) {
+        if (deepEqual(JSON.parse(JSON.stringify(arg)), JSON.parse(key))) {
+          return cache[key];
+        }
+      }
+
+      return (cache[JSON.stringify(arg)] = callback(...arg));
```

Comment:
if error appears in callback it's not write the result to the cache.
So let handle error of callback (use try/catch, write error message as value).